File: src/SimpleIndexVAO.cpp
#include "SimpleIndexVAO.h"
#include <iostream>
namespace ngl
{
SimpleIndexVAO::~SimpleIndexVAO()
{
  removeVAO();
}

void SimpleIndexVAO::draw() const
{
  if(m_allocated == false)
  {
    NGLMessage::addWarning("Warning trying to draw an unallocated VOA");
  }
  if(m_bound == false)
  {
    NGLMessage::addWarning("Warning trying to draw an unbound VOA");
  }
  glDrawElements(m_mode, static_cast< GLsizei >(m_indicesCount), m_indexType, static_cast< GLvoid * >(nullptr));
}

void SimpleIndexVAO::removeVAO()
{
  if(m_bound == true)
  {
    unbind();
  }
  if(m_allocated == true)
  {
    glDeleteBuffers(1, &m_buffer);
    glDeleteBuffers(1, &m_idxBuffer);
  }
  glDeleteVertexArrays(1, &m_id);
  m_allocated = false;
}

void SimpleIndexVAO::setData(const AbstractVAO::VertexData &_data)
{
  auto data = static_cast< const VertexData & >(_data);
  if(m_bound == false)
  {
    NGLMessage::addWarning("trying to set VOA data when unbound");
  }
  if(m_allocated == true)
  {
    glDeleteBuffers(1, &m_buffer);
  }

  glGenBuffers(1, &m_buffer);
  glGenBuffers(1, &m_idxBuffer);

  // now we will bind an array buffer to the first one and load the data for the verts
  glBindBuffer(GL_ARRAY_BUFFER, m_buffer);
  glBufferData(GL_ARRAY_BUFFER, static_cast< GLsizeiptr >(data.m_size), &data.m_data, data.m_mode);
  // we need to determine the size of the data type before we set it
  // in default to a ushort
  int size = sizeof(GLushort);
  switch(data.m_indexType)
  {
    case GL_UNSIGNED_INT:
      size = sizeof(GLuint);
      break;
    case GL_UNSIGNED_SHORT:
      size = sizeof(GLushort);
      break;
    case GL_UNSIGNED_BYTE:
      size = sizeof(GLubyte);
      break;
    default:
      NGLMessage::addWarning("wrong data type send for index value");
      break;
  }
  // now for the indices
  glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, m_idxBuffer);
  glBufferData(GL_ELEMENT_ARRAY_BUFFER, data.m_indexSize * static_cast< GLsizeiptr >(size), const_cast< GLvoid * >(data.m_indexData), data.m_mode);

  m_allocated = true;
  m_indexType = data.m_indexType;
}

#ifdef PYTHONBUILD
void SimpleIndexVAO::setData(size_t _size, const std::vector< Vec3 > &_data, unsigned int _indexSize, const std::vector< GLushort > &_indexData)
{
  if(m_bound == false)
  {
    NGLMessage::addWarning("trying to set VOA data when unbound");
  }
  if(m_allocated == true)
  {
    glDeleteBuffers(1, &m_buffer);
  }

  //    GLuint vboID;
  glGenBuffers(1, &m_buffer);

  //    GLuint iboID;
  glGenBuffers(1, &m_idxBuffer);

  // now we will bind an array buffer to the first one and load the data for the verts
  glBindBuffer(GL_ARRAY_BUFFER, m_buffer);
  glBufferData(GL_ARRAY_BUFFER, static_cast< GLsizeiptr >(_size), &_data[0].m_x, GL_STATIC_DRAW);
  // we need to determine the size of the data type before we set it
  // in default to a ushort
  int size = sizeof(GLushort);
  // now for the indices
  glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, m_idxBuffer);
  glBufferData(GL_ELEMENT_ARRAY_BUFFER, _indexSize * static_cast< GLsizeiptr >(size), &_indexData[0], GL_STATIC_DRAW);

  m_allocated = true;
  m_indexType = GL_UNSIGNED_SHORT;
}

#endif

Real *SimpleIndexVAO::mapBuffer([[maybe_unused]] unsigned int _index, GLenum _accessMode)
{
  Real *ptr = nullptr;
  bind();
  glBindBuffer(GL_ARRAY_BUFFER, m_buffer);
  ptr = static_cast< Real * >(glMapBuffer(GL_ARRAY_BUFFER, _accessMode));
  return ptr;
}

} // namespace ngl

File: py-ngl/src/ngl/simple_index_vao.py
import numpy as np
import OpenGL.GL as gl

from .abstract_vao import AbstractVAO, VertexData
from .log import logger


class IndexVertexData(VertexData):
    def __init__(self, data, size, indices, index_type, mode=gl.GL_STATIC_DRAW):
        super().__init__(data, size, mode)
        gl.GL_to_numpy_type = {
            gl.GL_UNSIGNED_INT: np.uint32,
            gl.GL_UNSIGNED_SHORT: np.uint16,
            gl.GL_UNSIGNED_BYTE: np.uint8,
        }
        numpy_dtype = gl.GL_to_numpy_type.get(index_type)
        if numpy_dtype is None:
            logger.error("SimpleIndexVAO: Unsupported index type")
            raise TypeError(f"Unsupported index type: {index_type}")

        self.indices = np.array(indices, dtype=numpy_dtype)
        self.index_type = index_type


class SimpleIndexVAO(AbstractVAO):
    def __init__(self, mode=gl.GL_TRIANGLES):
        super().__init__(mode)
        self.m_buffer = gl.glGenBuffers(1)
        self.m_idx_buffer = gl.glGenBuffers(1)
        self.m_index_type = gl.GL_UNSIGNED_INT

    def draw(self):
        if self.m_bound and self.m_allocated:
            gl.glDrawElements(self.m_mode, self.m_indicesCount, self.m_index_type, None)
        else:
            logger.error("SimpleIndexVAO not bound or not allocated")

    def set_data(self, data):
        if not isinstance(data, IndexVertexData):
            logger.error("SimpleIndexVAO: Unsupported index type")
            raise TypeError("data must be of type IndexVertexData")

        gl.glBindBuffer(gl.GL_ARRAY_BUFFER, self.m_buffer)
        gl.glBufferData(gl.GL_ARRAY_BUFFER, data.data.nbytes, data.data, data.mode)

        gl.glBindBuffer(gl.GL_ELEMENT_ARRAY_BUFFER, self.m_idx_buffer)
        gl.glBufferData(
            gl.GL_ELEMENT_ARRAY_BUFFER, data.indices.nbytes, data.indices, data.mode
        )

        self.m_allocated = True
        self.m_indicesCount = len(data.indices)
        self.m_index_type = data.index_type

    def remove_vao(self):
        gl.glDeleteBuffers(1, [self.m_buffer])
        gl.glDeleteBuffers(1, [self.m_idx_buffer])
        gl.glDeleteVertexArrays(1, [self.m_id])

    def get_buffer_id(self, index=0):
        return self.m_buffer

    def map_buffer(self, index=0, access_mode=gl.GL_READ_WRITE):
        gl.glBindBuffer(gl.GL_ARRAY_BUFFER, self.m_buffer)
        return gl.glMapBuffer(gl.GL_ARRAY_BUFFER, access_mode)
